crates/noirc_driver/src/lib.rs

```diff
@@ -183,23 +183,29 @@ impl Driver {
 
         // Create ABI for main function
         let func_meta = self.context.def_interner.function_meta(&main_function);
-        let abi = func_meta.into_abi(&self.context.def_interner);
+        let mut abi = func_meta.into_abi(&self.context.def_interner);
 
         let program = monomorphize(main_function, &self.context.def_interner);
 
         let blackbox_supported = acvm::default_is_black_box_supported(np_language.clone());
-        match create_circuit(program, np_language, blackbox_supported, show_ssa, show_output) {
-            Ok(circuit) => Ok(CompiledProgram { circuit, abi: Some(abi) }),
-            Err(err) => {
-                // The FileId here will be the file id of the file with the main file
-                // Errors will be shown at the call site without a stacktrace
-                let file = err.location.map(|loc| loc.file);
-                let files = &self.context.file_manager;
-                let error = reporter::report(files, &err.into(), file, allow_warnings);
-                reporter::finish_report(error as u32)?;
-                Err(ReportedError)
-            }
-        }
+        let (circuit, param_witnesses) =
+            match create_circuit(program, np_language, blackbox_supported, show_ssa, show_output) {
+                Ok((circuit, param_witnesses)) => (circuit, param_witnesses),
+                Err(err) => {
+                    // The FileId here will be the file id of the file with the main file
+                    // Errors will be shown at the call site without a stacktrace
+                    let file = err.location.map(|loc| loc.file);
+                    let files = &self.context.file_manager;
+                    let error = reporter::report(files, &err.into(), file, allow_warnings);
+                    reporter::finish_report(error as u32)?;
+                    return Err(ReportedError);
+                }
+            };
+
+        // TODO: Try to avoid this hack.
```

Comment:
Noting TODO

Reply:
I've changed `compile_circuit` to return the full ABI rather than stitching it together as we do here. I'm inclined to move `CompiledProgram` into `noirc_evaluator` and just return that directly.

Before I do, I want to check if  there's a reason we have this separation here.

Reply:
That's a good question -- the idea initially was that you could compile libraries and binaries(crates with main function), when its just a library, we returned None for Abi and when its a binary we returned the Abi with the circuit. 

I think this is irrelevant because if you have a library, you can only just do basic checks on it and won't produce a "library artifact" like you would a regular program. So we can put in in Evaluator and remove the optional I think